mmengine/optim/optimizer/default_constructor.py

```diff
@@ -301,8 +306,49 @@ def __call__(self, model: nn.Module) -> OptimWrapper:
             # set param-wise lr and weight decay recursively
             params: List = []
             self.add_params(params, model)
-            optimizer_cfg['params'] = params
+            # grouping parameters with the same hyper-parameters
+            if self.paramwise_cfg.get('reduce_param_groups', True):
+                optimizer_cfg['params'] = reduce_param_groups(params)
+            else:
+                optimizer_cfg['params'] = params
+            # enable foreach for pytorch 1.12.0+ to speed up training
+            if digit_version(TORCH_VERSION) >= digit_version('1.12.0'):
+                optimizer_cfg.setdefault('foreach', True)
+            else:
+                optimizer_cfg.pop('foreach', None)
+
             optimizer = OPTIMIZERS.build(optimizer_cfg)
         optim_wrapper = OPTIM_WRAPPERS.build(
             optim_wrapper_cfg, default_args=dict(optimizer=optimizer))
         return optim_wrapper
+
+
+def reduce_param_groups(params: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+    """Reorganize the parameter groups and merge duplicated groups. The number
+    of parameter groups needs to be as small as possible in order to
+    efficiently use the PyTorch multi-tensor optimizer. Therefore instead of
+    using a parameter_group per single parameter, we reorganize the parameter
+    groups and merge duplicated groups. This approach speeds up multi-tensor
+    optimizer significantly.
+
+    References: https://github.com/facebookresearch/detectron2/blob/main/detectron2/solver/build.py
```

Comment:
```suggestion
    Reference: https://github.com/facebookresearch/detectron2/blob/main/detectron2/solver/build.py
```

Reply:
I just follow the guide [here](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections):

![image](https://user-images.githubusercontent.com/57566630/230276125-b35781df-8ef3-484e-8eed-1e1ce565ce1b.png)